packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -70,217 +73,375 @@ fn get_pkg_path(name: &str, pkg_type: PackageType) -> String {
     fn import(
-        mv8: &'m MiniV8,
-        embedded: &Embedded<'m>,
+        scope: &mut v8::HandleScope<'s>,
+        context: v8::Local<'s, v8::Context>,
+        embedded: &Embedded<'s>,
         name: &str,
         pkg_type: PackageType,
-    ) -> Result<Self> {
+    ) -> JSRunnerResult<Self> {
         let path = get_pkg_path(name, pkg_type);
         tracing::debug!("Importing package from path `{}`", &path);
         let code = match fs::read_to_string(path.clone()) {
             Ok(s) => s,
             Err(_) => {
                 tracing::debug!("Couldn't read package file. It might intentionally not exist.");
                 // Packages don't have to use JS.
-                let fns = mv8.create_array();
-                fns.set(0, mv8::Value::Undefined)?;
-                fns.set(1, mv8::Value::Undefined)?;
-                fns.set(2, mv8::Value::Undefined)?;
+                let fns = v8::Array::new(scope, 3);
+                let undefined = v8::undefined(scope).into();
+                fns.set_index(scope, 0, undefined).ok_or_else(|| {
+                    JSRunnerError::V8("Could not create Undefined value".to_string())
+                })?;
+                fns.set_index(scope, 1, undefined).ok_or_else(|| {
+                    JSRunnerError::V8("Could not create Undefined value".to_string())
+                })?;
+                fns.set_index(scope, 2, undefined).ok_or_else(|| {
+                    JSRunnerError::V8("Could not create Undefined value".to_string())
+                })?;
+
                 return Ok(JsPackage { fns });
             }
         };
 
         // Avoid JS ReferenceError by wrapping potentially undeclared variables with `typeof`.
-        // (Double braces like `{{` are Rust's escape string for a single literal `{`.)
-        let wrapped_code = format!(
-            "((hash_util, hash_stdlib)=>{{{}
-            return [
-                typeof start_experiment === 'undefined' ? undefined : start_experiment,
-                typeof start_sim === 'undefined' ? undefined : start_sim,
-                typeof run_task === 'undefined' ? undefined : run_task
-            ]}})",
-            code
-        );
-        let pkg: mv8::Function<'_> = mv8
-            .eval(wrapped_code)
-            .map_err(|e| Error::PackageImport(path.clone(), e.into()))?;
-        let args = mv8::Values::from_vec(vec![
-            embedded.hash_util.clone(),
-            embedded.hash_stdlib.clone(),
-        ]);
-
-        let fns: mv8::Array<'_> = pkg
-            .call(args)
-            .map_err(|e| Error::PackageImport(path.clone(), e.into()))?;
-        if fns.len() != 3 {
-            return Err(Error::PackageImport(path.clone(), "Stray return".into()));
+        // Double braces like `{{` are Rust's escape string for a single literal `{`.
+        let wrapped_code = new_js_string(
+            scope,
+            &format!(
+                "((hash_util, hash_stdlib) => {{
+                    {code}
```

Comment:
`code` needs to be in the first line, otherwise, line numbers may be off by 1.

Reply:
Oh... that's too bad.
Can we at least keep a space between `{{` and `{code}`? I was pretty confused when reading it at first.